platform/pulp/plugin/changeset/main.py

```diff
@@ -0,0 +1,200 @@
+from collections import Sized, Iterable
+from gettext import gettext as _
+from logging import getLogger
+
+from django.db.utils import IntegrityError
+from django.db import transaction
+
+from pulp.app.models import ProgressBar, DownloadCatalog, RepositoryContent
+from pulp.download import Batch
+
+from ..tasking import Task
+
+from .iterator import DownloadIterator
+from .report import AddReport, DeleteReport
+
+
+log = getLogger(__name__)
+
+
+class ChangeSet:
+    """
+    A set of changes to be applied to a repository.
+
+    Attributes:
+        importer (pulp.plugin.Importer): An importer.
+        adds (SizedIterable): The content to be added to the repository.
+        deletes (SizedIterable): The content IDs to be removed.
+        deferred (bool): Downloading is deferred.  When true, downloads are not performed.
+
+    Examples:
+
+            >>>
+            >>> # TODO: Add example.
+            >>>
+    """
+
+    def __init__(self, importer, adds=(), deletes=()):
+        """
+        Args:
+            importer (pulp.plugin.Importer): An importer.
+            adds (SizedIterable): The content to be added to the repository.
+            deletes (SizedIterable): The content IDs to be removed.
+
+        Notes:
+            The content to be added may already exist but not be associated
+            to the repository. Existing content is fetched and used instead
+            of provided content models as needed.
+        """
+        self.importer = importer
+        self.adds = adds
+        self.deletes = deletes
+        self.deferred = False
+
+    def _associate(self, content):
+        """
+        Associate content.
+
+        Args:
+            content (RemoteContent): The content to be associated.
+        """
+        log.debug(_('Associate content.'))
+        try:
+            association = RepositoryContent(
+                repository=self.importer.repository,
+                content=content)
+            association.save()
+        except IntegrityError:
+            # Duplicate
+            pass
+
+    def _disassociate(self, content):
+        """
+        (Un)associate content with the repository.
+
+        Args:
+            content (pulp.plugin.Content): A content model to be unassociated.
+        """
+        log.debug(_('Un-associate content.'))
+        RepositoryContent.objects.delete(
+            repository=self.importer.repository,
+            content=content)
+        for artifact in content.artifacts:
+            DownloadCatalog.objects.delete(
+                importer=self.importer,
+                artifact=artifact)
```

Comment:
I wonder if it would be better to pass the list to the DB rather than do a python loop with individual DB calls.

``artifact__in=artifacts``

should do it. The only risk I guess would be if there are a CRAZY number of artifacts.